### PR TITLE
Sync rational-numbers; Don't use strict equality when comparing floats

### PR DESCRIPTION
**Pull request by angelikatyborska**
~This PR makes rational-numbers up to date with problem specs, but also addresses part of https://github.com/exercism/typescript/issues/1049 complaining about the `expreal` function tests failing for some students because of strict float comparisons.

The PR doesn't address the first part of that issue because I wasn't sure which way you would like to take that problem. I can push more commits to address that too if I get some guidance about which way you would like to go 🙂 ~

Resolves https://github.com/exercism/typescript/issues/1049

### CODE REVIEW

File: exercises/practice/rational-numbers/rational-numbers.test.ts [hunk 1]
```diff
@@ -150,17 +180,17 @@ describe('Exponentiation of a rational number', () => {
...
   xit('Raise a real number to a positive rational number', () => {
     const expected = 16.0
-    expect(new Rational(4, 3).expreal(8)).toEqual(expected)
+    expect(new Rational(4, 3).expreal(8)).toBeCloseTo(expected, 10)
   })
 
   xit('Raise a real number to a negative rational number', () => {
     const expected = 1.0 / 3.0
-    expect(new Rational(-1, 2).expreal(9)).toBeCloseTo(expected, 15)
+    expect(new Rational(-1, 2).expreal(9)).toBeCloseTo(expected, 10)
```
Comment (angelikatyborska) [inline exercises/practice/rational-numbers/rational-numbers.test.ts:188]: Turned the number of relevant digits down because on my machine this fails:
```js
expect(8**(4/3)).toBeCloseTo((8**(1/3))**4, 15)
```

Comment (SleeplessByte) [reply exercises/practice/rational-numbers/rational-numbers.test.ts:188]:
```text
15.999999999999998
16.000000000000000
```

Hehe.

**Comment by SleeplessByte** [comment]
@angelikatyborska I thought about this for the past few minutes.

Okay, so. `toEqual` in jest is basically a smart deep _property matcher_. Therefore, if, as the original issue describes, you create private properties (`Object.keys(instance)` does NOT include them), it's hidden. I don't like to rely on that at all. I can make non-hidden public properties that won't show up in `Object.keys` (aka are not enumerable). 

There are two solutions, I think, and you can pick which one you like:

1. Implement toString. Have people implement toString as `x/y`. I don't _love_ this because it's not standard notation, but it works. We then compare the `toString()` values everywhere instead. Unfortunately JavaScript doesn't have operator overloading :(
2. Force a public API of `rational.nominator` and `rational.denominator`:

```typescript
export class Rational {
  constructor(public readonly nominator = 0, public readonly denominator = 1) {
  }

  [...]
```

The above is shorthand for assigning to a public readonly property.
If we go with this, I would suggest a custom `assertSameRational(a, b)` or something like that, that lives in the test file, which compares the properties instead.

File: exercises/practice/rational-numbers/rational-numbers.test.ts [hunk 2]
```diff
@@ -1,197 +1,288 @@
...
+  expect(r1.denominator).toEqual(r2.denominator)
+}
+
+describe ('Constructor', () => {
+  it('Assigns public fields numerator and denominator', () => {
+    const rational = new Rational(5, 7)
+    expect(rational.numerator).toEqual(5)
+    expect(rational.denominator).toEqual(7)
+  })
+})
```
Comment (angelikatyborska) [inline exercises/practice/rational-numbers/rational-numbers.test.ts:14]: I had to add this extra test because, even requiring the fields to be public, leaving the fields empty or always equal to 0/1 would allow somebody to make all comparisons pass like the ticket https://github.com/exercism/typescript/issues/1049 mentions

**Comment by angelikatyborska** [comment]
I went for the second suggestion

File: exercises/practice/rational-numbers/rational-numbers.test.ts (hunk 2 above)
Comment (SleeplessByte) [reply exercises/practice/rational-numbers/rational-numbers.test.ts]: Yes, correct. Sorry should have listed that.

You can still fake it now by setting it to 5 and 7, always. Perhaps `assertSameRational` is changed to:

```typescript
assertSameRational(actual, 7, 6)
```

So at each stage we check the actual values?

If you want, I can write to code to be:

```typescript
expect(actual).toBeRational(7, 6)
```

Comment (angelikatyborska) [reply exercises/practice/rational-numbers/rational-numbers.test.ts]: Is this better https://github.com/exercism/typescript/pull/1277/commits/b6d4ebc1d2927862db897cbe38758c38b520e3f8 ?

File: exercises/practice/rational-numbers/.meta/proof.ci.ts [hunk 3]
```diff
@@ -49,10 +49,17 @@ export class Rational {
...
+      return new Rational(
+        Math.pow(this.numerator, n),
+        Math.pow(this.denominator, n)
+      )
+    } else {
+      return new Rational(
+        Math.pow(this.denominator, Math.abs(n)),
+        Math.pow(this.numerator, Math.abs(n))
+      )
+    }
```
Comment (SleeplessByte) [inline exercises/practice/rational-numbers/.meta/proof.ci.ts:62]: It's fine because it's a proof and not exemplar, but this would have worked too:

```typescript
return new Rational(
  Math.pow(this.denominator, Math.abs(n)),
  Math.pow(this.numerator, Math.abs(n))
)
```

The cost of the branch may, if this code was in the hot path, be as expensive as always removing the sign bit. 😁

**Review by SleeplessByte — APPROVED** [review]

File: exercises/practice/rational-numbers/rational-numbers.test.ts (hunk 2 above)
Comment (SleeplessByte) [reply exercises/practice/rational-numbers/rational-numbers.test.ts]: I think so, yeah! Hope you're also good with this.
